### ISSUE TITLE
Better :pred documentation

### ISSUE THREAD

**Issue by mz-pdm** [alphapapa/org-super-agenda#103]
It would be nice to clarify that the string contains org properties that can be extracted such as (I'm not sure it's the right way):
```
(get-text-property 0 'tags item)
```
It would help a lot those of us who are not that familiar with Org internals.

**Comment by alphapapa** [comment]
That could be helpful, but I think listing all the text properties would be 1) too much information, and 2) not necessarily up-to-date, since Org can do whatever it wants, as well as 3) time-consuming to produce such a list since there is no documented Org Agenda API (I had to figure it all out myself), and there are several Org Agenda commands that work differently, each possibly using their own subset of the possible properties.

If you want to investigate and compile a list of all the possible text properties and their value types, I'm willing to try to find a way to integrate that information into the documentation, but I think for most users, if they are one like yourself, they'll be able to find out the information they need anyway.  The documentation is already pretty complicated, and `:pred` is basically for advanced users.

What do you think?

**Comment by mz-pdm** [comment]
I didn't mean to list all the properties, just to clarify what kind of string :pred receives as the argument, i.e. an Org-propertized string. Adding a single example, such as the one above, would be sufficiently helpful. That single line easily explains that:
- The string is propertized.
- The desired properties are available at its beginning.
- One should use a non-keyword symbol to retrieve the desired property.

It would save org-super-agenda newcomers some debugging plus, ehm, as a bonus, Elisp documentation reading.

**Comment by alphapapa** [comment]
There is an example of using `:pred` here: https://github.com/alphapapa/org-super-agenda/blob/master/examples.org#by-predicate  I'd rather add to it than make the documentation more complicated.

However, what is your use case?  The example you gave of `(get-text-property 0 'tags item)` can already be accomplished with the `:tag` selector, and it can be combined with `:pred` using `:and`.  And if there are any useful text properties that aren't already usable with a built-in selector, we should probably add selectors for them rather than make users write predicates that call `get-text-property`.

In other words, I think using `:pred` is, or should be, the wrong way to match based on agenda text properties.  So if there's something missing in that regard, let's fix that.

**Comment by mz-pdm** [comment]
My use case is to identify mistyped tags. I use a fixed set of tags for todo entries and I want to identify todo entries with wrong tags. I use the following function in :pred to select those entries:
```
(defun my-org-agenda-unknown-tags-p (item)
  (seq-difference (get-text-property 0 'tags item) my-org-agenda-tags))
```
I think it is a typical example where :pred is actually useful.

**Comment by alphapapa** [comment]
Thanks, that's an interesting example!

You could also do it something like this:
```el
(setq org-super-agenda-groups
      `((:name "Unknown tags"
               :not (:tag (,@my-org-agenda-tags)))))
```
So the `:pred` selector isn't strictly necessary for that.  But it might sometimes be more convenient than backquoting/splicing the groups list.

When I get some time I'll improve the `:pred` example using the one you provided.  Thanks.

**Comment by mz-pdm** [comment]
Using `:not (:tag ...)` doesn't work since it excludes items with any of the given tags. An item with `:oktag:brokentag:` wouldn't be displayed.

**Comment by mz-pdm** [comment]
Good idea, works well this way.

**Comment by justinabrahms** [comment]
I came here looking for an issue quite like this.

The question I had was "What data type is `item` in `:pred`?"

Specifically, I'm hoping to get org-defined properties off of it, but I don't actually know what it is, much less which functions I can call on it.

**Comment by alphapapa** [comment]
The `:pred` selector is aimed at advanced users.  Using it will require some Elisp knowledge, and with that knowledge, one will be able to discover what's needed to use it.  For example, the example in the documentation shows that the predicate receives a string as its argument.  It would also be trivial to use a function which shows its argument, like:

```el
(let ((org-super-agenda-groups
       '((:pred (lambda (arg) (message "%s" arg))))))
  (org-agenda-list))
```
That's how I wrote the package in the first place, by exploring the Org Agenda code and its output.  

I don't want to explain details about how to write a predicate for the `:pred` selector, because that would require explaining how Org Agenda works internally, which is out-of-scope for this documentation.

And the more I explain, the deeper the questions will go.  If I explain that it receives a string with text properties, I'll be asked how to access them (answer: see the Emacs manual).  If I explain that, I'll be asked where the properties come from, and why certain ones aren't always present, and why they're of different types, etc (answer: see the Org Agenda source code--there is no other answer).  

The purpose of this project's documentation is not to explain how Org Agenda is implemented or how to write Elisp.  Like many things in Emacs and Org, some knowledge is prerequisite, knowledge which is out-of-scope for this project's documentation.

The ultimate answer is to read the source code, because it's Emacs.  :)

This is not to say that the documentation is perfect as-is.  Feel free to suggest specific improvements that would point users in the right direction.  But generally, pointers in the right direction is as much detail as I want to add about this topic.

**Comment by justinabrahms** [comment]
Thanks @alphapapa. I see the catch-22 you're in. Thanks to mz-pdm's notes, I found that it was a text-y thing that had properties. I attempted to log those properties, but they were truncated. I'm fairly green at elisp, as you might guess. Diving into the source I go. Thanks a bunch! :)

**Comment by alphapapa** [comment]
I wish I knew how much time I spent on this project waiting for Emacs to finish displaying very long lines of propertized text in the echo area (maybe Emacs 27 will help with that).  A few pointers:

* You can set `print-length` and `print-level` to avoid truncation.
* `data-debug-mode` may help with analyzing propertized strings.
* Edebug is a generally helpful tool.
* Every item in an agenda buffer is a propertized string that is passed through org-super-agenda.  You can see the text properties on them using `C-u C-x =`.
